atlasdb-client/src/main/java/com/palantir/atlasdb/cache/DefaultOffHeapCache.java

```diff
@@ -0,0 +1,178 @@
+import com.palantir.tritium.metrics.registry.MetricName;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+
+final class DefaultOffHeapCache<K, V> implements OffHeapCache<K, V> {
+    private static final Logger log = LoggerFactory.getLogger(DefaultOffHeapCache.class);
+    private static final String BATCHER_PURPOSE = "off-heap-cache";
+    private static final MetricName CACHE_HIT = constructCacheMetricName("cacheHit");
+    private static final MetricName CACHE_MISS = constructCacheMetricName("cacheMiss");
+    private static final MetricName CACHE_NUKE = constructCacheMetricName("cacheNuke");
+    private static final MetricName CACHE_SIZE = constructCacheMetricName("cacheSize");
+
+    private final LogicalPersistentStore<K, V> persistentStore;
+    private final LongSupplier maxSize;
+    private final AtomicReference<CacheDescriptor> cacheDescriptor = new AtomicReference<>();
+    private final TaggedMetricRegistry taggedMetricRegistry;
+    private final ConcurrentMap<K, V> inFlightRequests = new ConcurrentHashMap<>();
+    private final DisruptorAutobatcher<Map.Entry<K, V>, Map.Entry<K, V>> valuePutter;
+
+    DefaultOffHeapCache(
+            LogicalPersistentStore<K, V> persistentStore,
+            CacheDescriptor cacheDescriptor,
+            LongSupplier maxSize,
+            TaggedMetricRegistry taggedMetricRegistry) {
+        this.persistentStore = persistentStore;
+        this.cacheDescriptor.set(cacheDescriptor);
+        this.maxSize = maxSize;
+        this.taggedMetricRegistry = taggedMetricRegistry;
+        this.valuePutter = Autobatchers.coalescing(new WriteBatcher<>(this))
+                .safeLoggablePurpose(BATCHER_PURPOSE)
+                .build();
+        Gauge<Integer> cacheSizeGauge = () -> DefaultOffHeapCache.this.cacheDescriptor.get().currentSize().intValue();
+        taggedMetricRegistry.gauge(CACHE_SIZE, cacheSizeGauge);
+    }
+
+    public void clear() {
+        CacheDescriptor proposedCacheDescriptor = createNamespaceAndConstructCacheProposal(persistentStore);
+
+        CacheDescriptor previous = cacheDescriptor.getAndUpdate(prev -> proposedCacheDescriptor);
+        if (previous != null) {
+            persistentStore.dropNamespace(previous.storeNamespace());
+        }
+    }
+
+    @Override
+    public void put(K key, V value) {
+        if (inFlightRequests.putIfAbsent(key, value) != null) {
+            return;
+        }
+        Futures.getUnchecked(valuePutter.apply(Maps.immutableEntry(key, value)));
+    }
+
+    @Override
+    public Optional<V> getIfPresent(K key) {
+        Optional<V> value = getValue(key);
+
+        if (value.isPresent()) {
+            taggedMetricRegistry.meter(CACHE_HIT).mark();
+        } else {
+            taggedMetricRegistry.meter(CACHE_MISS).mark();
+        }
+        return value;
+    }
+
+    private Optional<V> getValue(K key) {
+        V inFlight = inFlightRequests.get(key);
```

Comment:
use canonical `Optional` form for this entire method

Reply:
plus is this safe? there's no guarantee it was written no? is this not assuming the immutability of transactions table?

Reply:
This is removed.